atlasdb-dbkvs/src/main/java/com/palantir/atlasdb/keyvalue/dbkvs/impl/DbKvs.java

```diff
@@ -658,6 +638,176 @@ public int compare(Entry<Cell, Long> entry1, Entry<Cell, Long> entry2) {
+                    @Override
+                    public List<Map.Entry<Cell, Value>> apply(DbReadTable table) {
+                        return Iterables.getOnlyElement(extractRowColumnRangePage(table, table.getRowsColumnRange(rowList, ts, range), rowList).values());
+                    }
+                });
+                if (nextPage.isEmpty()) {
+                    return SimpleTokenBackedResultsPage.create(startCol, ImmutableList.<Entry<Cell, Value>>of(), false);
+                }
+                byte[] lastCol = nextPage.get(nextPage.size() - 1).getKey().getColumnName();
+                if (isEndOfColumnRange(lastCol, columnRangeSelection.getEndCol())) {
+                    return SimpleTokenBackedResultsPage.create(lastCol, nextPage, false);
+                }
+                byte[] nextCol = RangeRequests.nextLexicographicName(lastCol);
+                return SimpleTokenBackedResultsPage.create(nextCol, nextPage, true);
+            }
+
+        }.iterator());
+    }
+
+    private boolean isEndOfColumnRange(byte[] lastCol, byte[] endCol) {
+        return RangeRequests.isLastRowName(lastCol) || Arrays.equals(RangeRequests.nextLexicographicName(lastCol), endCol);
+    }
+
+    private Map<byte[], List<Map.Entry<Cell, Value>>> getFirstRowsColumnRangePage(TableReference tableRef, List<byte[]> rows,
+                                                          ColumnRangeSelection columnRangeSelection, long ts) {
+        Stopwatch watch = Stopwatch.createStarted();
+        try {
+            return runRead(tableRef, new Function<DbReadTable, Map<byte[], List<Map.Entry<Cell, Value>>>>() {
+                @Override
+                public Map<byte[], List<Map.Entry<Cell, Value>>> apply(DbReadTable table) {
+                    return extractRowColumnRangePage(table, table.getRowsColumnRange(rows, ts, columnRangeSelection), rows);
+                }
+            });
+        } finally {
+            log.info("Call to KVS.getFirstRowColumnRangePage on table {} took {} ms.", tableRef, watch.elapsed(TimeUnit.MILLISECONDS));
+        }
+    }
+
+    private Map<byte[], List<Map.Entry<Cell, Value>>> extractRowColumnRangePage(DbReadTable table,
+                                                                              ClosableIterator<AgnosticLightResultRow> iter,
+                                                                              List<byte[]> rows) {
+        Map<Sha256Hash, byte[]> hashesToBytes = Maps.newHashMapWithExpectedSize(rows.size());
+        Map<Sha256Hash, List<Cell>> cellsByRow = Maps.newHashMap();
+        for (byte[] row : rows) {
+            Sha256Hash rowHash = Sha256Hash.computeHash(row);
+            hashesToBytes.put(rowHash, row);
+            cellsByRow.put(rowHash, Lists.newArrayList());
+        }
+
+        boolean hasOverflow = table.hasOverflowValues();
+        Map<Cell, Value> values = Maps.newHashMap();
+        Map<Cell, OverflowValue> overflowValues = Maps.newHashMap();
+
+        try {
```

Comment:
Should probably move try to top in case one of the above fails so we cleanup and close properly